File: backend/app/api/v1/endpoints/circulation.py
"""
Circulation management endpoints
"""
from fastapi import APIRouter, HTTPException, status, Depends, Query, Response
from typing import Optional
from ....models.circulation import (
    CirculationCreate,
    CirculationReturn,
    CirculationUpdate,
    CirculationResponse,
    CirculationDetailResponse,
    CirculationListResponse,
    CirculationStatsResponse
)
from ....services.circulation_service import CirculationService
import csv
import io

router = APIRouter()


def get_circulation_service() -> CirculationService:
    """Dependency to get circulation service instance"""
    return CirculationService()


@router.get("", response_model=CirculationListResponse, summary="Get all circulation records")
async def get_circulation_records(
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    search: Optional[str] = Query(None, description="Search by user name, book title"),
    status: Optional[str] = Query(None, description="Filter by status (active, overdue, returned, reserved)"),
    user_type: Optional[str] = Query(None, description="Filter by user type"),
    due_date_filter: Optional[str] = Query(None, description="Filter by due date (today, tomorrow, week, overdue)"),
    sort_by: str = Query("issue_date", description="Sort field"),
    sort_order: str = Query("desc", regex="^(asc|desc)$", description="Sort order"),
    circulation_service: CirculationService = Depends(get_circulation_service)
):
    """
    Get paginated list of circulation records with optional filtering and sorting

    - **page**: Page number (default: 1)
    - **page_size**: Items per page (default: 20, max: 100)
    - **search**: Search in user name, book title
    - **status**: Filter by status (active, overdue, returned, reserved)
    - **user_type**: Filter by user type
    - **due_date_filter**: Filter by due date (today, tomorrow, week, overdue)
    - **sort_by**: Sort field (issue_date, due_date, etc.)
    - **sort_order**: Sort order (asc or desc)
    """
    try:
        result = await circulation_service.get_circulation_records(
            page=page,
            page_size=page_size,
            search=search,
            status=status,
            user_type=user_type,
            due_date_filter=due_date_filter,
            sort_by=sort_by,
            sort_order=sort_order
        )
        return result
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch circulation records: {str(e)}"
        )


@router.get("/stats", response_model=CirculationStatsResponse, summary="Get circulation statistics")
async def get_circulation_stats(circulation_service: CirculationService = Depends(get_circulation_service)):
    """
    Get circulation statistics including:
    - Active issues
    - Overdue books
    - Returned today
    - Reserved books
    - Total fines
    - Total fines paid
    - Average borrow duration
    - Most borrowed books
    - Most active users
    """
    try:
        stats = await circulation_service.get_circulation_stats()
        return stats
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch circulation stats: {str(e)}"
        )


@router.get("/export", summary="Export circulation records to CSV")
async def export_circulation_records(
    search: Optional[str] = Query(None, description="Search by user name, book title"),
    status: Optional[str] = Query(None, description="Filter by status"),
    user_type: Optional[str] = Query(None, description="Filter by user type"),
    due_date_filter: Optional[str] = Query(None, description="Filter by due date"),
    circulation_service: CirculationService = Depends(get_circulation_service)
):
    """
    Export circulation records to CSV file with optional filtering
    """
    try:
        # Get all records with filters (no pagination)
        result = await circulation_service.get_circulation_records(
            page=1,
            page_size=10000,  # Large number to get all records
            search=search,
            status=status,
            user_type=user_type,
            due_date_filter=due_date_filter,
            sort_by="issue_date",
            sort_order="desc"
        )

        # Create CSV in memory
        output = io.StringIO()
        writer = csv.writer(output)

        # Write header
        writer.writerow([
            'ID', 'User Name', 'User Role', 'Book Title', 'Category',
            'Shelf Location', 'Issue Date', 'Due Date', 'Return Date',
            'Status', 'Days Left/Overdue', 'Book Condition', 'Fine Amount (OMR)',
            'Fine Paid', 'Notes'
        ])

        # Write data
        for record in result['items']:
            writer.writerow([
                record['id'],
                record['user_name'],
                record['user_role'],
                record['book_title'],
                record.get('category', ''),
                record.get('shelf_location', ''),
                record['issue_date'],
                record['due_date'],
                record.get('return_date', ''),
                record['status'],
                f"{record['days_left']} days" if record['days_left'] >= 0 else f"{abs(record['days_left'])} days overdue",
                record.get('book_condition', ''),
                record.get('fine_amount', 0),
                'Yes' if record.get('fine_paid') else 'No',
                record.get('notes', '')
            ])

        # Return CSV response
        output.seek(0)
        return Response(
            content=output.getvalue(),
            media_type="text/csv",
            headers={
                "Content-Disposition": f"attachment; filename=circulation_records_export.csv"
            }
        )

    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to export circulation records: {str(e)}"
        )


@router.get("/{record_id}", response_model=CirculationDetailResponse, summary="Get circulation record by ID")
async def get_circulation_record(
    record_id: str,
    circulation_service: CirculationService = Depends(get_circulation_service)
):
    """
    Get single circulation record by ID with detailed information
    """
    try:
        record = await circulation_service.get_circulation_record(record_id)
        if not record:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Circulation record not found"
            )
        return record
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch circulation record: {str(e)}"
        )


@router.post("", response_model=CirculationResponse, status_code=status.HTTP_201_CREATED, summary="Issue a book")
async def issue_book(
    circulation_data: CirculationCreate,
    circulation_service: CirculationService = Depends(get_circulation_service)
):
    """
    Issue a book to a user (create new circulation record)

    - **user_id**: User UUID
    - **book_id**: Book UUID
    - **issue_date**: Issue date (defaults to today)
    - **due_date**: Due date (typically 15 days from issue)
    - **send_email**: Send email notification (default: false)
    - **print_receipt**: Print receipt (default: false)
    - **notes**: Optional notes
    """
    try:
        record = await circulation_service.issue_book(circulation_data.dict())
        return record
    except Exception as e:
        error_msg = str(e)
        if "not found" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User or book not found"
            )
        if "already borrowed" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Book is already borrowed"
            )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to issue book: {error_msg}"
        )


@router.post("/{record_id}/return", response_model=CirculationResponse, summary="Return a book")
async def return_book(
    record_id: str,
    return_data: CirculationReturn,
    circulation_service: CirculationService = Depends(get_circulation_service)
):
    """
    Process book return

    - **return_date**: Return date (defaults to today)
    - **book_condition**: Book condition (good, fair, damaged)
    - **notes**: Optional notes about condition or issues
    """
    try:
        record = await circulation_service.return_book(record_id, return_data.dict())
        return record
    except Exception as e:
        error_msg = str(e)
        if "not found" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Circulation record not found"
            )
        if "already returned" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Book has already been returned"
            )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to process return: {error_msg}"
        )


@router.patch("/{record_id}", response_model=CirculationResponse, summary="Update circulation record")
async def update_circulation_record(
    record_id: str,
    update_data: CirculationUpdate,
    circulation_service: CirculationService = Depends(get_circulation_service)
):
    """
    Update existing circulation record

    All fields are optional - only provided fields will be updated
    """
    try:
        # Filter out None values
        data = {k: v for k, v in update_data.dict().items() if v is not None}

        if not data:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="No fields to update"
            )

        record = await circulation_service.update_circulation_record(record_id, data)
        return record
    except HTTPException:
        raise
    except Exception as e:
        error_msg = str(e)
        if "not found" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Circulation record not found"
            )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update circulation record: {error_msg}"
        )


@router.delete("/{record_id}", status_code=status.HTTP_204_NO_CONTENT, summary="Delete circulation record")
async def delete_circulation_record(
    record_id: str,
    circulation_service: CirculationService = Depends(get_circulation_service)
):
    """
    Delete circulation record by ID
    """
    try:
        await circulation_service.delete_circulation_record(record_id)
        return None
    except Exception as e:
        error_msg = str(e)
        if "not found" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Circulation record not found"
            )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete circulation record: {error_msg}"
        )

File: frontend/lib/api/circulation.ts
/**
 * Circulation API Service
 *
 * Circulation management API endpoints (aligned with backend implementation)
 * Backend endpoints: /api/v1/circulation
 */

import apiClient from './client';

// ============================================================================
// Type Definitions (matching backend models)
// ============================================================================

export interface CirculationRecord {
  id: string;
  user_id: string;
  book_id: string;
  user_name: string;
  user_role: string;
  book_title: string;
  category?: string;
  shelf_location?: string;
  issue_date: string;
  due_date: string;
  return_date?: string;
  status: 'active' | 'overdue' | 'returned' | 'reserved';
  days_left: number;
  book_condition?: string;
  fine_amount?: number;
  fine_paid?: boolean;
  notes?: string;
}

export interface CirculationStats {
  active_issues: number;
  overdue_books: number;
  returned_today: number;
  reserved_books: number;
  total_fines: number;
  total_fines_paid: number;
  average_borrow_duration: number;
  most_borrowed_books: Array<{
    book_id: string;
    title: string;
    count: number;
  }>;
  most_active_users: Array<{
    user_id: string;
    name: string;
    count: number;
  }>;
}

export interface CirculationFilters {
  page?: number;
  page_size?: number;
  search?: string;
  status?: 'active' | 'overdue' | 'returned' | 'reserved';
  user_type?: string;
  due_date_filter?: 'today' | 'tomorrow' | 'week' | 'overdue';
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface IssueBookRequest {
  user_id: string;
  book_id: string;
  issue_date?: string;
  due_date: string;
  send_email?: boolean;
  print_receipt?: boolean;
  notes?: string;
}

export interface ReturnBookRequest {
  return_date?: string;
  book_condition?: 'good' | 'fair' | 'damaged';
  notes?: string;
}

export interface UpdateCirculationRequest {
  due_date?: string;
  notes?: string;
  fine_amount?: number;
  fine_paid?: boolean;
}

export interface PaginatedCirculationResponse {
  items: CirculationRecord[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// ============================================================================
// Circulation Operations (matching backend /circulation endpoints)
// ============================================================================

/**
 * Get all circulation records with filters and pagination
 * Backend: GET /circulation
 */
export const getCirculationRecords = async (
  filters?: CirculationFilters
): Promise<PaginatedCirculationResponse> => {
  const response = await apiClient.get<PaginatedCirculationResponse>('/circulation', {
    params: filters,
  });
  return response.data;
};

/**
 * Get circulation record by ID
 * Backend: GET /circulation/{record_id}
 */
export const getCirculationRecord = async (recordId: string): Promise<CirculationRecord> => {
  const response = await apiClient.get<CirculationRecord>(`/circulation/${recordId}`);
  return response.data;
};

/**
 * Issue a book (create circulation record)
 * Backend: POST /circulation
 */
export const issueBook = async (data: IssueBookRequest): Promise<CirculationRecord> => {
  const response = await apiClient.post<CirculationRecord>('/circulation', data);
  return response.data;
};

/**
 * Return a book
 * Backend: POST /circulation/{record_id}/return
 */
export const returnBook = async (
  recordId: string,
  data: ReturnBookRequest
): Promise<CirculationRecord> => {
  const response = await apiClient.post<CirculationRecord>(
    `/circulation/${recordId}/return`,
    data
  );
  return response.data;
};

/**
 * Update circulation record
 * Backend: PATCH /circulation/{record_id}
 */
export const updateCirculationRecord = async (
  recordId: string,
  data: UpdateCirculationRequest
): Promise<CirculationRecord> => {
  const response = await apiClient.patch<CirculationRecord>(
    `/circulation/${recordId}`,
    data
  );
  return response.data;
};

/**
 * Delete circulation record
 * Backend: DELETE /circulation/{record_id}
 */
export const deleteCirculationRecord = async (recordId: string): Promise<void> => {
  await apiClient.delete(`/circulation/${recordId}`);
};

/**
 * Get circulation statistics
 * Backend: GET /circulation/stats
 */
export const getCirculationStats = async (): Promise<CirculationStats> => {
  const response = await apiClient.get<CirculationStats>('/circulation/stats');
  return response.data;
};

/**
 * Export circulation records to CSV
 * Backend: GET /circulation/export
 */
export const exportCirculationRecords = async (
  filters?: Omit<CirculationFilters, 'page' | 'page_size'>
): Promise<Blob> => {
  const response = await apiClient.get('/circulation/export', {
    params: filters,
    responseType: 'blob',
  });
  return response.data;
};

// ============================================================================
// Convenience Functions
// ============================================================================

/**
 * Get active loans (circulation records with status=active)
 */
export const getActiveLoans = async (
  page: number = 1,
  pageSize: number = 20
): Promise<PaginatedCirculationResponse> => {
  return getCirculationRecords({
    page,
    page_size: pageSize,
    status: 'active',
  });
};

/**
 * Get overdue books
 */
export const getOverdueBooks = async (
  page: number = 1,
  pageSize: number = 20
): Promise<PaginatedCirculationResponse> => {
  return getCirculationRecords({
    page,
    page_size: pageSize,
    status: 'overdue',
  });
};

/**
 * Get circulation records for a specific user
 */
export const getUserCirculationRecords = async (
  userId: string,
  page: number = 1,
  pageSize: number = 20
): Promise<PaginatedCirculationResponse> => {
  return getCirculationRecords({
    page,
    page_size: pageSize,
    search: userId, // Backend will search by user_id in the search field
  });
};

/**
 * Quick checkout - issue book with default settings
 */
export const quickCheckout = async (
  userId: string,
  bookId: string,
  daysToLoan: number = 15
): Promise<CirculationRecord> => {
  const dueDate = new Date();
  dueDate.setDate(dueDate.getDate() + daysToLoan);

  return issueBook({
    user_id: userId,
    book_id: bookId,
    due_date: dueDate.toISOString(),
  });
};

/**
 * Quick return - return book in good condition
 */
export const quickReturn = async (recordId: string): Promise<CirculationRecord> => {
  return returnBook(recordId, {
    book_condition: 'good',
  });
};

// Export all for convenience
export default {
  getCirculationRecords,
  getCirculationRecord,
  issueBook,
  returnBook,
  updateCirculationRecord,
  deleteCirculationRecord,
  getCirculationStats,
  exportCirculationRecords,
  getActiveLoans,
  getOverdueBooks,
  getUserCirculationRecords,
  quickCheckout,
  quickReturn,
};
